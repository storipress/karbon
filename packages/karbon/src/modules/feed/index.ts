import { addServerHandler, createResolver, defineNuxtModule } from '@nuxt/kit'
export type ModuleOptions = boolean

export default defineNuxtModule({
  meta: {
    name: 'karbon/feed',
    configKey: 'feed',
  },
  defaults: {},
  setup(_) {
    const { resolve } = createResolver(import.meta.url)

    addServerHandler({
      route: '/atom.xml',
      handler: resolve('./runtime/routes/atom.xml'),
    })
  },
})
