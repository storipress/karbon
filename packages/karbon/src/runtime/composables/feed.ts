import type { MaybeRef } from '@vueuse/core'
import { toValue } from '@vueuse/core'
import { joinURL } from 'ufo'
import type { Desk } from './page-meta'
import { useHead, useRuntimeConfig } from '#imports'

export function useFeedLink() {
  const runtimeConfig = useRuntimeConfig()
  const siteUrl = runtimeConfig?.public?.siteUrl || ''

  useHead({
    link: [
      {
        key: 'karbon-atom-feed',
        rel: 'alternate',
        href: joinURL(siteUrl, '/atom.xml'),
        type: 'application/atom+xml',
      },
    ],
  })
}

export function useDeskFeedLink(deskMeta: MaybeRef<Desk>) {
  const runtimeConfig = useRuntimeConfig()
  const siteUrl = runtimeConfig?.public?.siteUrl || ''

  useHead(() => ({
    link: [
      {
        key: 'karbon-atom-feed',
        rel: 'alternate',
        href: joinURL(siteUrl, `/atom/${toValue(deskMeta).slug}.xml`),
        type: 'application/atom+xml',
        tagPriority: 'high',
      },
    ],
  }))
}
