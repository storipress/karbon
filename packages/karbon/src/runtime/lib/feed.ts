import { Feed } from 'feed'
import { joinURL, withQuery, withTrailingSlash, withoutTrailingSlash } from 'ufo'
import { XMLBuilder, XMLParser } from 'fast-xml-parser'

export function createFeed({
  siteUrl,
  siteName,
  siteDescription,
  feedUrl,
}: {
  siteUrl: string
  siteName: string
  siteDescription: string
  feedUrl: string
}) {
  return new Feed({
    id: withTrailingSlash(siteUrl),
    link: withTrailingSlash(siteUrl),
    title: siteName,
    description: siteDescription,
    updated: new Date(),
    feedLinks: {
      atom: joinURL(siteUrl, feedUrl),
    },
    copyright: `Â© ${siteName} ${new Date().getFullYear()} All Rights Reserved`,
  })
}

export function addFeedPageLinks(atomXml: string, siteUrl: string, currentPage: number, maxPage: number) {
  const option = {
    ignoreAttributes: false,
    attributeNamePrefix: '@_',
    cdataPropName: '__cdata',
    format: true,
  }
  const parser = new XMLParser(option)
  const builder = new XMLBuilder(option)

  const atomJson = parser.parse(atomXml)

  const rssUrl = `${withoutTrailingSlash(siteUrl)}/atom.xml`
  const previousLink =
    currentPage > 1 ? [{ '@_rel': 'previous', '@_href': withQuery(rssUrl, { page: currentPage - 1 }) }] : []
  const nextLink =
    currentPage < maxPage ? [{ '@_rel': 'next', '@_href': withQuery(rssUrl, { page: currentPage + 1 }) }] : []

  const buildAtomXml = builder.build({
    ...atomJson,
    feed: {
      ...atomJson.feed,
      link: [...atomJson.feed.link, ...previousLink, ...nextLink],
    },
  })
  return buildAtomXml
}
