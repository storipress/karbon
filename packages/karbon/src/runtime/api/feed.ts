import { gql } from '@apollo/client/core/index.js'
import { useStoripressClient } from '../composables/storipress-client'
import { getAllWithPagination } from './helper'
import type { RawArticleLike } from './normalize-article'
import { normalizeArticle } from './normalize-article'

const ListArticles = gql`
  query ListArticles($page: Int!, $desk: ID, $desk_ids: [ID!]) {
    articles(page: $page, desk: $desk, desk_ids: $desk_ids, sortBy: [{ column: UPDATED_AT, order: DESC }]) {
      paginatorInfo {
        count
        lastPage
        hasMorePages
      }
      data {
        id
        title
        slug
        sid
        published_at
        html
        plaintext
      }
    }
  }
`

const GetDesk = gql`
  query GetDesk($slug: String) {
    desk(slug: $slug) {
      id
      sid
      name
      slug
      seo
      order
      open_access
      desks {
        id
        name
        slug
        seo
      }
    }
  }
`

export function listFeedArticles(filter?: { desk: string; tag: string; author: string; desk_ids: string }) {
  return getAllWithPagination(ListArticles, filter, ({ articles: { paginatorInfo, data } }) => {
    const res = data.map((data: RawArticleLike) => normalizeArticle(data))
    return {
      paginatorInfo,
      data: res,
    }
  })
}

export async function getDeskWithSlug(slug: string) {
  const client = useStoripressClient()
  const { data } = await client.query({ query: GetDesk, variables: { slug } })

  return data.desk
}
