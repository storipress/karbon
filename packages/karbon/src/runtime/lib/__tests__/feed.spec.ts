import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest'
import { addFeedPageLinks, createFeed, generateAtomFeed } from '../feed'

beforeEach(() => {
  vi.useFakeTimers()
  vi.setSystemTime(new Date('2024-04-24T00:00:00.000Z'))
})

afterEach(() => {
  vi.useRealTimers()
})

describe('createFeed', () => {
  it('can create feed with default config', () => {
    const feed = createFeed({
      siteName: 'site name',
      siteDescription: 'site description',
      siteUrl: 'https://example.com',
      feedUrl: 'atom.xml',
    })

    feed.addItem({
      title: 'title',
      date: new Date('2024-04-24T00:00:00.000Z'),
      link: 'https://example.com/post',
    })

    expect(feed.atom1()).toMatchSnapshot()
  })
})

describe('generateAtomFeed', () => {
  it('can generate atom feed', () => {
    const atom = generateAtomFeed({
      articles: [
        {
          id: '1',
          slug: 'slug',
          authors: [{ name: 'author' }] as any,
          title: 'title',
          html: 'html',
          plaintext: 'plaintext',
          updated_at: '2024-04-24T00:00:00.000Z',
          published_at: '2024-04-24T00:00:00.000Z',
        },
      ],
      getArticleURL: () => '/posts/slug',
      siteName: 'site name',
      siteDescription: 'site description',
      siteUrl: 'https://example.com',
      feedUrl: 'atom.xml',
    })

    expect(atom).toMatchSnapshot()
  })
})

const ATOM_FIXTURE = `
<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <id>https://example.com/</id>
    <title>site name</title>
    <updated>2024-04-24T00:00:00.000Z</updated>
    <generator>https://github.com/jpmonette/feed</generator>
    <link rel="alternate" href="https://example.com/"/>
    <link rel="self" href="https://example.com/atom.xml"/>
    <subtitle>site description</subtitle>
    <rights>Â© site name 2024 All Rights Reserved</rights>
    <entry>
        <title type="html"><![CDATA[title]]></title>
        <id>https://example.com/post</id>
        <link href="https://example.com/post"/>
        <updated>2024-04-24T00:00:00.000Z</updated>
    </entry>
</feed>
`

describe('addFeedPageLinks', () => {
  it('can add feed page links', () => {
    expect(addFeedPageLinks(ATOM_FIXTURE, 'https://example.com', 2, 3)).toMatchSnapshot()
  })
})
