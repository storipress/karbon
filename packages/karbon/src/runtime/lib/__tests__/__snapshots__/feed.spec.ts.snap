// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`addFeedPageLinks > can add feed page links 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
  <id>https://example.com/</id>
  <title>site name</title>
  <updated>2024-04-24T00:00:00.000Z</updated>
  <generator>https://github.com/jpmonette/feed</generator>
  <link rel="alternate" href="https://example.com/"></link>
  <link rel="self" href="https://example.com/atom.xml"></link>
  <link rel="previous" href="https://example.com/atom.xml?page=1"></link>
  <link rel="next" href="https://example.com/atom.xml?page=3"></link>
  <subtitle>site description</subtitle>
  <rights>© site name 2024 All Rights Reserved</rights>
  <entry>
    <title type="html">
      <![CDATA[title]]>
    </title>
    <id>https://example.com/post</id>
    <link href="https://example.com/post"></link>
    <updated>2024-04-24T00:00:00.000Z</updated>
  </entry>
</feed>
"
`;

exports[`createFeed > can create feed with default config 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <id>https://example.com/</id>
    <title>site name</title>
    <updated>2024-04-24T00:00:00.000Z</updated>
    <generator>https://github.com/jpmonette/feed</generator>
    <link rel="alternate" href="https://example.com/"/>
    <link rel="self" href="https://example.com/atom.xml"/>
    <subtitle>site description</subtitle>
    <rights>© site name 2024 All Rights Reserved</rights>
    <entry>
        <title type="html"><![CDATA[title]]></title>
        <id>https://example.com/post</id>
        <link href="https://example.com/post"/>
        <updated>2024-04-24T00:00:00.000Z</updated>
    </entry>
</feed>"
`;

exports[`generateAtomFeed > can generate atom feed 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">
    <id>https://example.com/</id>
    <title>site name</title>
    <updated>2024-04-24T00:00:00.000Z</updated>
    <generator>https://github.com/jpmonette/feed</generator>
    <link rel="alternate" href="https://example.com/"/>
    <link rel="self" href="https://example.com/atom.xml"/>
    <subtitle>site description</subtitle>
    <rights>© site name 2024 All Rights Reserved</rights>
    <entry>
        <title type="html"><![CDATA[title]]></title>
        <id>https://example.com/posts/slug</id>
        <link href="https://example.com/posts/slug"/>
        <updated>2024-04-24T00:00:00.000Z</updated>
        <summary type="html"><![CDATA[plaintext]]></summary>
        <content type="html"><![CDATA[html]]></content>
        <author>
            <name>author</name>
        </author>
        <published>2024-04-24T00:00:00.000Z</published>
    </entry>
</feed>"
`;
